libraries/standard/http/src/http_client.c

```diff
@@ -867,51 +1011,64 @@ static HTTPStatus_t getFinalResponseStatus( HTTPParsingState_t parsingState,
+                                                 HTTPResponse_t * pResponse,
+                                                 const HTTPRequestHeaders_t * pRequestHeaders )
 {
     HTTPStatus_t returnStatus = HTTP_SUCCESS;
     size_t totalReceived = 0u;
     size_t currentReceived = 0u;
     HTTPParsingContext_t parsingContext = { 0 };
-    uint8_t shouldRecv = 0u;
+    uint8_t shouldRecv = 1u;
+    uint8_t isHeadResponse = 0u;
 
     assert( pTransport != NULL );
     assert( pTransport->recv != NULL );
     assert( pResponse != NULL );
+    assert( pRequestHeaders != NULL );
+    assert( pRequestHeaders->headersLen >= HTTP_MINIMUM_REQUEST_LINE_LENGTH );
 
-    /* Initialize the parsing context. */
-    returnStatus = HTTPClient_InitializeParsingContext( &parsingContext,
-                                                        pResponse->pHeaderParsingCallback );
-
-    if( returnStatus == HTTP_SUCCESS )
+    /* The parsing context needs to know if the response is for a HEAD request.
+     * The third-party parser requires parsing is manually indicated to stop
+     * in the httpParserOnHeadersCompleteCallback() for a HEAD response,
+     * otherwise the parser will not indicate the message was complete. */
+    if( strncmp( ( const char * ) ( pRequestHeaders->pBuffer ),
+                 HTTP_METHOD_HEAD,
+                 sizeof( HTTP_METHOD_HEAD ) - 1u ) == 0 )
     {
-        shouldRecv = 1u;
+        isHeadResponse = 1u;
     }
 
+    /* Initialize the parsing context for parsing the response received from the
+     * network. */
+    initializeParsingContextForFirstResponse( &parsingContext );
+
     while( shouldRecv == 1u )
     {
         /* Receive the HTTP response data into the pResponse->pBuffer. */
-        returnStatus = receiveHttpResponse( pTransport,
-                                            pResponse->pBuffer + totalReceived,
-                                            pResponse->bufferLen - totalReceived,
-                                            &currentReceived );
+        returnStatus = receiveHttpData( pTransport,
+                                        pResponse->pBuffer + totalReceived,
+                                        pResponse->bufferLen - totalReceived,
+                                        &currentReceived );
 
         if( returnStatus == HTTP_SUCCESS )
         {
-            if( currentReceived > 0u )
-            {
-                totalReceived += currentReceived;
-                /* Data is received into the buffer and must be parsed. */
-                returnStatus = HTTPClient_ParseResponse( &parsingContext,
-                                                         pResponse->pBuffer + totalReceived,
-                                                         currentReceived );
-            }
+            /* Data is received into the buffer and must be parsed. Parsing is
+             * invoked even with a length of zero. A length of zero indicates to
```

Comment:
Where are we invoking it with a length of zero?

Reply:
To signal an EOF to the http-parser library.